doc/sg_read_block_limits.8
.TH SG_READ_BLOCK_LIMITS "8" "November 2022" "sg3_utils\-1.48" SG3_UTILS
.SH NAME
sg_read_block_limits \- send SCSI READ BLOCK LIMITS command
.SH SYNOPSIS
.B sg_read_block_limits
[\fI\-\-help\fR] [\fI\-\-hex\fR] [\fI\-\-mloi\fR] [\fI\-\-raw\fR]
[\fI\-\-readonly\fR] [\fI\-\-verbose\fR] [\fI\-\-version\fR]
\fIDEVICE\fR
.SH DESCRIPTION
.\" Add any additional description here
Send a SCSI READ BLOCK LIMITS command to \fIDEVICE\fR and outputs the
response. This command is defined for tape (drives) and its description
is found in the SSC documents at https://www.t10.org .
.SH OPTIONS
Arguments to long options are mandatory for short options as well.
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-H\fR, \fB\-\-hex\fR
output response in hex (rather than decode it).
.TP
\fB\-m\fR, \fB\-\-mloi\fR
sets the MLOI bit in the READ BLOCK LIMITS command and if that
succeeds, prints out the Maximum Logical Object Identifier (MLOI)
value. The MLOI bit was introduced in the ssc4r02.pdf draft.
.TP
\fB\-r\fR, \fB\-\-raw\fR
output response in binary to stdout.
.TP
\fB\-R\fR, \fB\-\-readonly\fR
open \fIDEVICE\fR in read\-only mode. The default is to open it in
read\-write mode.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the level of verbosity, (i.e. debug output).
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.SH EXIT STATUS
The exit status of sg_read_block_limits is 0 when it is successful. Otherwise
see the sg3_utils(8) man page.
.SH EXAMPLES
It is usually okay to use no options. Here is an invocation (on the first
line following the "#" command prompt) followed by some typical output:
.PP
   # sg_read_block_limits /dev/st0
   Read Block Limits results:
     Minimum block size: 1 byte(s)
     Maximum block size: 16777215 byte(s), 16383 KB, 15 MB
     Granularity: 0
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2009\-2022 Douglas Gilbert
.br
This software is distributed under a BSD\-2\-Clause license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B sg3_utils(sg3_utils)


ChangeLog
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for pre-release sg3_utils-1.49 [20231027] [svn: r1060]
  - JSON: make output more consistent so most command
    responses have a *_paramter_data or similar sub-object
  - apply https://github.com/doug-gilbert/sg3_utils/pull/39
    and its revision [20230807] mainly for Android
  - sg_inq: update version descriptors to T10 table 20230814
  - sg_wr_mode: fix --contents and --cfile= handling
  - sg_vpd: inform about SPC-6 (sg_inq knew already)
    https://github.com/doug-gilbert/sg3_utils/issues/40
  - sg_inq+sg_vpd: show depopulation time in hours and
    minutes as well as its raw form: seconds
    - add RSCS field to Block limits ext. vpage [sbc5r5]
  - sg_turs: fix missing CDB on several codepaths
  - sg_requests: improve buffer truncation
  - sg_logs: correct tape Device statisitics lpage [0x14]
    serial number parameters [0x40 and 0x41] per
    https://github.com/doug-gilbert/sg3_utils/issues/45
  - sg_get_elem_status: various improvements
  - sg_stream_ctl: add --hex, --inhex=, --json= and --raw
    options; additions in sbc5r05
  - rescan-scsi-bus.sh: fix multipath resize without update 
    https://github.com/doug-gilbert/sg3_utils/pull/43.diff
    and 44.diff
  - new error code: SG_LIB_PROGRESS_NOT_READY for long
    duration 'in progress' additional sense codes
  - lib/sg_pt_solaris.c : fix compilation issue per
    https://github.com/doug-gilbert/sg3_utils/pull/41
    - add --mfile=MF option for decoding mask in the MF file
  - lib/sgj_haj_*() functions: tweak hex_haj argument meaning
  - scripts/00-scsi-sg3_config.rules : apply
    https://github.com/doug-gilbert/sg3_utils/pull/42
  - testing/sg_chk_inq_vd.c: test internal table against T10
    version descriptor file

Changelog for released sg3_utils-1.48 [20230801] [svn: r1042]
  - decoding utilities: add --json[=JO] and --js-file=JFN
    options, short form changed to -j[=JO]
  - sg_sat_datetime: new utility for setting or accessing
    the data and time on a ATA device using a SAT layer.
  - sg_z_act_query: new utility for sending either a
    Zone activate or Zone query command
  - sg_rep_density: new utility for decoding the response of
    Report density support command [ssc (tape)]
  - sg_rem_rest_elem: new utility for removing or restoring
    elements
  - sg_write_attr: new utility, Write Attribute is a SPC
    command that is used mainly by tape drives (ssc)
  - --hex option cleanup, especially in sg_inq and sg_vpd
    - swap meaning of '-H' and '-HH' in sg_ses
  - sg_rtpg: https://github.com/hreinecke/sg3_utils/pull/79
    applied
  - netbsd: experimental port, scsi only
  - rescan-scsi-bus.sh: with '-r' it crashed due to change in
    rev 867 in sg_inq.c from Device_type= to PDT= .
    Change script to use either
    - undo regression in rev 815 that added newline after
      each LUN in the default (no option) output
    - rev 815 changed the order of listing hosts from
      numeric to alphabetical, change it back to numeric
    - https://github.com/doug-gilbert/sg3_utils/pull/17
      applied with tweaks: add timeout parameter
    - clean $norm handling
    - fix handling of '-I <secs>' option
    - sgdevice26: do not traverse sg class if scsi_device
      is not added
    - add -no-lip-scan option
    - https://github.com/doug-gilbert/sg3_utils/pull/21
    - speed testonline() when peripheral_qualifier != 0
      see https://github.com/doug-gilbert/sg3_utils/issues/24
    - improve handling of USB SD card readers (with no cards
      present)
    - speed multipath scans with many LUNs, cache multipath
      LUN info in temporary file, see
      https://github.com/doug-gilbert/sg3_utils/issues/22
  - udev rules: restrict use of ambiguous device IDs patchset
    (3 part) from Martin Wilck applied
  - sg_rep_zones: add Report zone starting LBA granularity
    field in REPORT ZONES response [zbc2r12]
    - add --brief option, show part of header and last
      descriptor fetched
    - add --find ZT option to find the first occurrence of
      ZT; if ZT prefixed by - or ! find first not equal to ZT
    - add --statistics option
  - sg_get_elem_status: change '--maxlen=' option default to
    1056 (header plus 32 physical element status descriptors)
    - add current (maximum) number of depopulated elements
      fields
  - sg_get_lba_status: decode new LBA accessibility field
  - sg_decode_sense: add --nodecode option
  - sg_logs: tweak the meaning of --list option to more closely
    reflect the contents of log pages 0x0 and 0x0,0xff
    - make '-lll' set union of log pages 0x0 and 0x0,0xff
    - add --exclude option to exclude vendor specific pages
      and parameters
    - add --undefined option for hex format of undefined/
      unrecognized fields
    - for short binary fields, remove address (index) from
      the left hand side of each line of hex
    - improve 'last_n' log pages; supply VPD and mode pages
      with their name (if T10 defined)
    - update names for TapeAlert lpage
    - allow --page selection with --inhex=FN
  - sg_modes: improve handling of zbc disks with pdt=0x14
    - add --multiple option for all available page controls
  - sg_inq, sg_vpd: merge VPD page processing for T10
    defined pages, VS pages still differ
    - add Capacity/Product identification mapping VPD page
    - Device Identication VPD page, change
      "IEEE Company_id" to "AOI" as per spc6r06.pdf
    - add support for Hitachi/HP open-v ldev names
  - sg_inq: add (many) version descriptors as per spc6r08
    - add --quiet option, like the one sg_vpd has
  - sg_vpd: apply github pull 18 (missing LF)
    - add --sinq_inraw=RFN option
  - sg_opcodes: cleanup error reporting
    - add --inhex=FN to process earlier -HHH
  - sg_format: allow disk formats on ZBC (zoned) disks

src/sg_read_block_limits.c
/*
 * Copyright (c) 2009-2018 Douglas Gilbert.
 * All rights reserved.
 * Use of this source code is governed by a BSD-style
 * license that can be found in the BSD_LICENSE file.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdbool.h>
#include <string.h>
#include <getopt.h>
#define __STDC_FORMAT_MACROS 1
#include <inttypes.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "sg_lib.h"
#include "sg_cmds_basic.h"
#include "sg_cmds_extra.h"
#include "sg_unaligned.h"
#include "sg_pr2serr.h"

/* A utility program originally written for the Linux OS SCSI subsystem.
 *
 *
 * This program issues the SCSI READ BLOCK LIMITS command (SSC) to the given
 * SCSI device.
 */

static const char * version_str = "1.08 20180219";

#define MAX_READ_BLOCK_LIMITS_LEN 6

static uint8_t readBlkLmtBuff[MAX_READ_BLOCK_LIMITS_LEN];


static struct option long_options[] = {
        {"help", no_argument, 0, 'h'},
        {"hex", no_argument, 0, 'H'},
        {"raw", no_argument, 0, 'r'},
        {"readonly", no_argument, 0, 'R'},
        {"verbose", no_argument, 0, 'v'},
        {"version", no_argument, 0, 'V'},
        {0, 0, 0, 0},
};


static void
usage()
{
    pr2serr("Usage: sg_read_block_limits  [--help] [--hex] [--raw] "
            "[--readonly]\n"
            "                             [--verbose] [--version] "
            "DEVICE\n"
            "  where:\n"
            "    --help|-h          print out usage message\n"
            "    --hex|-H           output response in hexadecimal\n"
            "    --raw|-r           output response in binary to stdout\n"
            "    --readonly|-R      open DEVICE in read-only mode\n"
            "    --verbose|-v       increase verbosity\n"
            "    --version|-V       print version string and exit\n\n"
            "Performs a SCSI READ BLOCK LIMITS command and decode the "
            "response\n"
            );
}

static void
dStrRaw(const char * str, int len)
{
    int k;

    for (k = 0; k < len; ++k)
        printf("%c", str[k]);
}

int
main(int argc, char * argv[])
{
    bool do_raw = false;
    bool readonly = false;
    bool verbose_given = false;
    bool version_given = false;
    int sg_fd, k, m, res, c;
    int do_hex = 0;
    int verbose = 0;
    int ret = 0;
    uint32_t max_block_size;
    uint16_t min_block_size;
    uint8_t granularity;
    const char * device_name = NULL;

    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "hHrRvV", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'h':
        case '?':
            usage();
            return 0;
        case 'H':
            ++do_hex;
            break;
        case 'r':
            do_raw = true;
            break;
        case 'R':
            readonly = true;
            break;
        case 'v':
            verbose_given = true;
            ++verbose;
            break;
        case 'V':
            version_given = true;
            break;
        default:
            pr2serr("invalid option -%c ??\n", c);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if (NULL == device_name) {
            device_name = argv[optind];
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                pr2serr("Unexpected extra argument: %s\n", argv[optind]);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
#ifdef DEBUG
    pr2serr("In DEBUG mode, ");
    if (verbose_given && version_given) {
        pr2serr("but override: '-vV' given, zero verbose and continue\n");
        verbose_given = false;
        version_given = false;
        verbose = 0;
    } else if (! verbose_given) {
        pr2serr("set '-vv'\n");
        verbose = 2;
    } else
        pr2serr("keep verbose=%d\n", verbose);
#else
    if (verbose_given && version_given)
        pr2serr("Not in DEBUG mode, so '-vV' has no special action\n");
#endif
    if (version_given) {
        pr2serr("version: %s\n", version_str);
        return 0;
    }

    if (NULL == device_name) {
        pr2serr("missing device name!\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }

    sg_fd = sg_cmds_open_device(device_name, readonly, verbose);
    if (sg_fd < 0) {
        if (verbose)
            pr2serr("open error: %s: %s\n", device_name,
                    safe_strerror(-sg_fd));
        ret = sg_convert_errno(-sg_fd);
        goto the_end2;
    }

    memset(readBlkLmtBuff, 0x0, 6);
    res = sg_ll_read_block_limits(sg_fd, readBlkLmtBuff, 6, true, verbose);
    ret = res;
    if (0 == res) {
      if (do_hex) {
        hex2stdout(readBlkLmtBuff, sizeof(readBlkLmtBuff), 1);
        goto the_end;
      } else if (do_raw) {
        dStrRaw((const char *)readBlkLmtBuff, sizeof(readBlkLmtBuff));
        goto the_end;
      }

      max_block_size = sg_get_unaligned_be32(readBlkLmtBuff + 0);
      // first byte contains granularity field
      granularity = (max_block_size >> 24) & 0x1F;
      max_block_size = max_block_size & 0xFFFFFF;
      min_block_size = sg_get_unaligned_be16(readBlkLmtBuff + 4);
      k = min_block_size / 1024;
      pr2serr("Read Block Limits results:\n");
      pr2serr("\tMinimum block size: %u byte(s)",
              (unsigned int)min_block_size);
      if (k != 0)
        pr2serr(", %d KB", k);
      pr2serr("\n");
      k = max_block_size / 1024;
      m = max_block_size / 1048576;
      pr2serr("\tMaximum block size: %u byte(s)",
              (unsigned int)max_block_size);
      if (k != 0)
        pr2serr(", %d KB", k);
      if (m != 0)
        pr2serr(", %d MB", m);
      pr2serr("\n");
      pr2serr("\tGranularity: %u",
              (unsigned int)granularity);
      pr2serr("\n");
    } else {
        char b[80];

        sg_get_category_sense_str(res, sizeof(b), b, verbose);
        pr2serr("Read block limits: %s\n", b);
        if (0 == verbose)
            pr2serr("    try '-v' option for more information\n");
    }

the_end:
    res = sg_cmds_close_device(sg_fd);
    if (res < 0) {
        pr2serr("close error: %s\n", safe_strerror(-res));
        if (0 == ret)
            ret = sg_convert_errno(-res);
    }
the_end2:
    if (0 == verbose) {
        if (! sg_if_can2stderr("sg_read_block_limits failed: ", ret))
            pr2serr("Some error occurred, try again with '-v' or '-vv' for "
                    "more information\n");
    }
    return (ret >= 0) ? ret : SG_LIB_CAT_OTHER;
}
